File: matrices/0773-sliding-puzzle.py
from collections import deque
from typing import List
import unittest

# https://leetcode.com/problems/sliding-puzzle/

# python3 -m unittest matrices/0773-sliding-puzzle.py

class Solution(unittest.TestCase):
    # # Approach: Depth-First Search
    # # Time: O((m⋅n)!×(m⋅n)^2)
    # # Space: O((m⋅n)!)
    # def slidingPuzzle(self, board: List[List[int]]) -> int:
    #     neighbors = [[1,3],[0,2,4],[1,5],[0,4],[1,3,5],[2,4]]
    #     dist = {}
    #     def dfs(state: List[str], curr: int, moves: int) -> None:
    #         key = "".join(state)
    #         if key in dist and dist[key] <= moves:
    #             return
    #         dist[key] = moves
    #         for next in neighbors[curr]:
    #             state[curr], state[next] = state[next], state[curr]
    #             dfs(state, next, moves+1)
    #             state[curr], state[next] = state[next], state[curr]
    #     state = [str(val) for row in board for val in row]
    #     dfs(state, state.index('0'), 0)
    #     return dist.get("123450", -1)

    # Approach: Breadth-First Search
    # Time: O((m⋅n)!×(m⋅n))
    # Space: O((m⋅n)!)
    def slidingPuzzle(self, board: List[List[int]]) -> int:
        neighbors = [[1,3],[0,2,4],[1,5],[0,4],[1,3,5],[2,4]]
        state = "".join([str(val) for row in board for val in row])
        queue = deque([(state, state.index("0"), 0)])
        visited = set([state])
        while queue:
            state, curr, length = queue.popleft()
            if state == "123450":
                return length
            b = list(state)
            for next in neighbors[curr]:
                b[curr], b[next] = b[next], b[curr]
                new_state = "".join(b)
                if new_state not in visited:
                    queue.append((new_state, next, length+1))
                    visited.add(new_state)
                b[curr], b[next] = b[next], b[curr]
        return -1

    def test(self):
        for board, expected in [
            ([[1,2,3],[4,0,5]], 1),
            ([[1,2,3],[5,4,0]], -1),
            ([[4,1,2],[5,0,3]], 5),
            ([[1,3,4],[0,2,5]], 14),
            ([[3,2,4],[1,5,0]], 14),
            ([[2,4,1],[5,3,0]], 12),
            ([[0,5,2],[4,3,1]], 15),
            ([[4,2,0],[5,1,3]], 7),
            ([[3,0,1],[4,2,5]], -1),
            ([[3,2,1],[4,0,5]], -1),
            ([[1,2,3],[4,5,0]], 0),
        ]:
            output = self.slidingPuzzle(board)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: matrices/0773-sliding-puzzle.go
package matrices

import (
	"bytes"
)

// https://leetcode.com/problems/sliding-puzzle/

// Approach: Breadth-First Search
// Time: O((m⋅n)!×(m⋅n))
// Space: O((m⋅n)!)
func slidingPuzzle(board [][]int) int {
	neighbors := [][]int{{1, 3}, {0, 2, 4}, {1, 5}, {0, 4}, {1, 3, 5}, {2, 4}}
	state := make([]byte, 0, 6)

	for _, row := range board {
		for _, cell := range row {
			state = append(state, byte('0'+cell))
		}
	}

	queue := []string{string(state)}
	seen := map[string]bool{string(state): true}
	length := 0

	for size := len(queue); size > 0; size = len(queue) {
		for idx := 0; idx < size; idx++ {
			state = []byte(queue[idx])
			if string(state) == "123450" {
				return length
			}

			curr := bytes.IndexByte(state, '0')
			for _, next := range neighbors[curr] {
				state[curr], state[next] = state[next], state[curr]
				if nextState := string(state); !seen[nextState] {
					queue = append(queue, nextState)
					seen[nextState] = true
				}

				state[curr], state[next] = state[next], state[curr]
			}
		}

		queue = queue[size:]
		length++
	}

	return -1
}

// // Approach: Depth-First Search
// // Time: O((m⋅n)!×(m⋅n)^2)
// // Space: O((m⋅n)!)
// func slidingPuzzle(board [][]int) int {
// 	var state = make([]byte, 0, 6)
// 	for _, row := range board {
// 		for _, cell := range row {
// 			state = append(state, byte('0'+cell))
// 		}
// 	}
// 	var neighbors = [6][]int{{1, 3}, {0, 2, 4}, {1, 5}, {0, 4}, {1, 3, 5}, {2, 4}}
// 	var dist = make(map[string]int)
// 	var dfs func(state []byte, curr int, moves int)
// 	dfs = func(state []byte, curr int, moves int) {
// 		var key = string(state)
// 		if d, ok := dist[key]; ok && d <= moves {
// 			return
// 		}
// 		dist[key] = moves
// 		for _, next := range neighbors[curr] {
// 			state[curr], state[next] = state[next], state[curr]
// 			dfs(state, next, moves+1)
// 			state[curr], state[next] = state[next], state[curr]
// 		}
// 	}
// 	dfs(state, bytes.IndexByte(state, '0'), 0)
// 	if d, ok := dist["123450"]; ok {
// 		return d
// 	}
// 	return -1
// }
